File: Strivers A2Z DSA course/2. Learn Important Sorting Techniques/2.2. Sorting-II/RecursiveInsertionSort.java
public class RecursiveInsertionSort {

    public void insertionSort(int arr[], int n) {
        if (n <= 1) {
            return;
        }
        insertionSort(arr, n - 1);
        int j = n - 1;
        while (j > 0 && (arr[j - 1] > arr[j])) {
            int temp = arr[j];
            arr[j] = arr[j - 1];
            arr[j - 1] = temp;
            j--;
        }
    }

}

File: Strivers A2Z DSA course/2. Learn Important Sorting Techniques/2.2. Sorting-II/Recursive Insertion Sort.cpp
// C program for insertion sort
#include <stdio.h>
#include <math.h>

/* Function to print an array */
void printArray(int arr[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}


// } Driver Code Ends
class Solution
{

     public:
    //Function to sort the array using insertion sort algorithm.
    void insertionSort(int arr[], int n)
    {
        if(n <= 1){
            return;
        }
        // as insertion sort starts from i=1 and go from right to left in second loop so we need to bring j=1;
        insertionSort(arr, n-1);
        int j = n-1;
        while(j>0 && (arr[j-1] > arr[j])){
            int temp = arr[j];
            arr[j] = arr[j-1];
            arr[j-1] = temp;
            j--;
        }
        
        
    }
};

//{ Driver Code Starts.
int main()
{
    int arr[1000],n,T,i;

    scanf("%d",&T);

    while(T--){

    scanf("%d",&n);

    for(i=0;i<n;i++)
      scanf("%d",&arr[i]);

    Solution ob;
    ob.insertionSort(arr, n);
    printArray(arr, n);
    }
    return 0;
}

// } Driver Code Ends